hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -588,6 +590,8 @@ public void onCompactionCompleted(RocksDB db,
             return;
           }
 
+          waitForTarballCreation();
```

Comment:
Waiting like this is probably not safe.  The doc for the onCompactionCompleted() listeners advises against doing it:

"Note that this function must be implemented in a way such that it should not run for an extended period of time before the function returns. Otherwise, RocksDB may be blocked."

https://javadoc.io/static/org.rocksdb/rocksdbjni/7.3.1/org/rocksdb/AbstractEventListener.html#onCompactionCompleted-org.rocksdb.RocksDB-org.rocksdb.CompactionJobInfo-

Reply:
Maybe q the requests and only write them when the lock is taken, outside of the listener?

Reply:
Maybe there is a better way.  The transaction logs have the seq number in the name, right?

if we could start a new compaction log when the tarball creation starts,  then tarball creation process could look at the file names and skip copying over logs whose name indicate a transaction number that is too large for the checkpoint in this tarball.  Would that work?

Reply:
Just transaction seq number in log doesn't solve the problem. I tired to explain it here: https://github.com/apache/ozone/commit/8db58df12318108b2d4e3cf8dc9bfeaccc4fadbe#diff-1d5c6c892a2d24d85ca002dafcb9f39341192baaa7b5323d2325b5fc7a238c7f
 and wrote a coding using ThreadLocal: https://github.com/apache/ozone/commit/8db58df12318108b2d4e3cf8dc9bfeaccc4fadbe#diff-1d5c6c892a2d24d85ca002dafcb9f39341192baaa7b5323d2325b5fc7a238c7fR500
 
 But problem with that assumption  that CompactionBeginListener and CompactionCompletedListener are running on a same thread is wrong.

Reply:
@GeorgeJahad In this case its is OK to block on compactionCompleted() because that's the intent. We do not want compactions to be completed unless tarball creation is completed.

Reply:
My understanding is that rocksdb itself will block, not just the compactions, but I'm not sure.  Do you know differently?

Reply:
Yup, this should not be a problem.

Reply:
For now it looks like this would at most block RocksDB compaction worker threads. It is unclear if it would indeed block other RocksDB operations (column family reads and writes).

Reply:
https://github.com/facebook/rocksdb/wiki/RocksDB-Overview#avoiding-stalls
This indicates that compaction threads are also used for flushing memtables, and if they get backed up, writes will stall, (which is what I've been afraid of.)

It also indicates that you can workaround the issue by configuring a dedicated flush threadpool.  But I still trying to think of a better solution.

Reply:
@GeorgeJahad I agree we should try a better solution.

What I've been thinking is that, since the original intent of taking a DB checkpoint (even before Ozone snapshot feature is implemented) is to transfer a **consistent point-in-time** copy of the active DB to the bootstrapping follower, we could apply the same principle to the _extras_ we added for the Ozone snapshot feature, namely:

1. Compaction log
2. SST backup directory

So an ideal solution would be: if RocksDB had `onCheckpointBegin` (or `onCheckpointCompleted`) listener, we could take a filesystem snapshot for both and include that point-in-time copy of compaction log and SST backup dir in the tarball instead. Based on the assumption that RocksDB won't perform checkpointing while another compaction thread is running (not verified yet).

Problems are:
1. Only some relatively modern FSes (ZFS, btrfs, [UFS](https://en.wikipedia.org/wiki/Unix_File_System), APFS) supports snapshots, and commonly-used `ext4` doesn't. So this won't be a portable solution. -- That is unless we enable this only on supported FSes. Otherwise we resort to copying / hard-linking instead.
2. Latest RocksDB doesn't have checkpoint event listeners, we could add them if we want to tho: https://javadoc.io/static/org.rocksdb/rocksdbjni/8.3.2/org/rocksdb/AbstractEventListener.html#i0

Reply:
I like the idea of copying the compaction logs and hard linking the compaction backup data.

I'm just not convinced we need a checkpoint listener.  After thinking about it and talking to @prashantpogde , I think the compaction listener is the right place.  I just think we are pausing for too long.

Right now, we pause until the network transfer to the follower completes.  That seems dangerously long.  But waiting long enough to copy the compaction logs, make the backup hard links, and do the checkpoint, seems reasonable.

Reply:
Using the compaction listener should also work but not as ideal.

Because in this case tarball creation process (which includes DB checkpointing) does not know if or when a compaction will happen, tarball creation thread has to block compaction threads similar to what we already have in the listener while we add existing compaction log to the tarball (no need to copy it elsewhere, just stream it into the tarball?) and hard-link the backup SST files somewhere. -- It just occurred to me that we might need to block DAG cleanup thread as well to prevent backup SST files from being cleaned up while tarball is being created? cc @hemantk-12 

Another solution without blocking or adding explicit lock would be: if tarball creation is in-progress, the compaction listener can append to a **staging** compaction log and hard-link SSTs to a **staging** SST backup dir, and some other threads will have to be responsible of merging the staging parts back. Tarball creation process and OM startup also has to trigger a check to merge them if there were any previous staging logs and backup SSTs.

Reply:
>  we might need to block DAG cleanup thread as well to prevent backup SST files from being cleaned up while tarball is being created?

I believe I'm already handling this:
https://github.com/apache/ozone/blob/master/hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java#L1172-L1177

https://github.com/apache/ozone/blob/master/hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java#L1484-L1488

Is there something else you think is needed?